### ISSUE TITLE
JSON rpc interface broke after upgrading to the plugin system

### ISSUE THREAD

**Issue by renepickhardt** [ElementsProject/lightning#2135]
### Issue and Steps to Reproduce
after upgrading to the plugin system which supports JSON API calls the pylightning json rpc interface breaks with a call like `rpc_interface.getinfo()`
the main reason seems that the response is shorter than 1024 Bytes and the socket does not close anymore. 

c.f.: 
```
  File "test.py", line 9, in <module>
    res = rpc_interface.getinfo()
  File "/Users/rpickhardt/anaconda3/lib/python3.6/site-packages/lightning/lightning.py", line 275, in getinfo
    return self.call("getinfo")
  File "/Users/rpickhardt/anaconda3/lib/python3.6/site-packages/lightning/lightning.py", line 75, in call
    resp = self._readobj(sock)
  File "/Users/rpickhardt/anaconda3/lib/python3.6/site-packages/lightning/lightning.py", line 32, in _readobj
    b = sock.recv(1024)
```

I was able to quickfix this by changing https://github.com/ElementsProject/lightning/blob/master/contrib/pylightning/lightning/lightning.py#L34
to `b = sock.recv(1024)``

I do not provide this patch as a pull request because I want to look at the changes to the server side. I realized that https://github.com/ElementsProject/lightning/blob/master/lightningd/lightningd.c#L705 is new. Before there was `setup_jsonrpc(ld, ld->rpc_filename);` instead of `jsonrpc_listen(ld->jsonrpc, ld);` so I guess the way the socket is handled within c-lightning has changed and breaks the RPC interface.

on a sidenote:
Shouldn't it be possible from a plugin to send json messages directly to `lightningd` without the necessity to specify the rpc-interface? Generally the plugin does not know where the data directory of the lightning node that calls the plugin is placed and would not know where the RPC interface is placed. since the plugin already talks to lightningd over stdout and stdin I believe there should be a way of making regular rpc-calls

**Comment by conscott** [comment]
Can confirm this issue. I was in the process of making a ticket before @renepickhardt beat me to it!

**Comment by renepickhardt** [comment]
@conscott I am currently comparing: 

void jsonrpc_listen(struct jsonrpc *jsonrpc, struct lightningd *ld)
https://github.com/ElementsProject/lightning/blob/d7e94a93866c457ed327c8b1b3ef37015a2aed73/lightningd/jsonrpc.c#L768

with 
void setup_jsonrpc(struct lightningd *ld, const char *rpc_filename)
https://github.com/ElementsProject/lightning/blob/ad1cf8b40c203444519d30284743a43b786744c0/lightningd/jsonrpc.c#L663

as I believe the error should be in there and fixing it there should be better than doing this quickfix on the pylightning file. 

I guess this issue should also be for @cdecker

**Comment by renepickhardt** [comment]
hmmm besides this assertion at https://github.com/ElementsProject/lightning/blob/d7e94a93866c457ed327c8b1b3ef37015a2aed73/lightningd/jsonrpc.c#L775
and a line flip of the log entry as well as the reversed arguments and new function name the functions are the same. So I guess I have to continue looking through the diff to see what has changed with how the socket and responses are handled

**Comment by renepickhardt** [comment]
found the mistake! if you look at the diff of this commit: 
https://github.com/ElementsProject/lightning/commit/b2378654d70aea2a5b456f948eb67104dd50a5fe#diff-1550ef6e9533217d6251a66fa023a861

the API was changed to have a double `\n\n` at the end of the json command for easier parsing. However the python file was still running with the old API of just one `\n`

so the real patch will be to fix: 

```
               if buff[-3:] != b' }\n':                                                                                                                                                                                                                                                              
                     continue     
```
in pylightning with
```
               if buff[-4:] != b' }\n\n':                                                                                                                                                                                                                                                              
                     continue     
```

however looking at: 
https://github.com/ElementsProject/lightning/blob/d7e94a93866c457ed327c8b1b3ef37015a2aed73/contrib/pylightning/lightning/lightning.py#L31
it seems to be already updated. 

Looking at pipy https://pypi.org/project/pylightning/#files it seems that the file was updated only on August 20th 2018. 

so there won't even be the necessity for a pull request

**Comment by cdecker** [comment]
Oh, good catch, will push a new version to pipy :+1:

**Comment by cdecker** [comment]
Version 0.0.5 is up: https://pypi.org/manage/project/pylightning/release/0.0.5/
```
pip install -U pylightning
```

Should fix your problems ^^

**Comment by renepickhardt** [comment]
This is still ugly. If people with old nodes now install pylightning the calls will break. There should be a way to fix the pylightning version to the clightning version or be backwards compatible

**Comment by cdecker** [comment]
Yeah, agreed, but we don't really have a nice way of doing this. This is why we added the `\n\n` in the first place.

**Comment by rustyrussell** [comment]
No, we revert the patch that makes pylightning rely on this, until the standard 1-version depreciation window...
